pandas/core/indexes/base.py

```diff
@@ -1680,23 +1680,29 @@ def append(self, other):
         names = set([obj.name for obj in to_concat])
         name = None if len(names) > 1 else self.name
 
-        if self.is_categorical():
-            # if calling index is category, don't check dtype of others
-            from pandas.core.indexes.category import CategoricalIndex
-            return CategoricalIndex._append_same_dtype(self, to_concat, name)
+        res = self._concat(to_concat)
 
-        typs = _concat.get_dtype_kinds(to_concat)
+        res.name = name
+        return res
 
+    @classmethod
+    def _concat(self, to_concat):
```

Comment:
and then you have the same exact problem in that the code lives in multiple unrelaled places. Please use the pattern I suggest.